In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import LabelEncoder, StandardScaler
df =  pd.read_csv("C:/Users/madhu/Downloads/FinMan/df_modelling.csv")
desc = pd.read_csv('C:/Users/madhu/Downloads/FinMan/Train_data_dictionary.csv')

### Basic Understanding of the Data

In [2]:
desc

Variable                                         Definition
0                        ID                        Unique Identifier for a row
1                 City_Code                 Code for the City of the customers
2               Region_Code               Code for the Region of the customers
3         Accomodation_Type                   Customer Owns or Rents the house
4       Reco_Insurance_Type  Joint or Individual type for the recommended i...
5                 Upper_Age                       Maximum age of the customer 
6                Lower _Age                        Minimum age of the customer
7                 Is_Spouse         If the customers are married to each other
8                       NaN                      (in case of joint insurance) 
9          Health_Indicator          Encoded values for health of the customer
10  Holding_Policy_Duration  Duration (in years) of holding policy (a polic...
11      Holding_Policy_Type                             Type of holding policy
12          Reco_Policy_Cat     Encoded value for recommended health insurance
13      Reco_Policy_Premium  Annual Premium (INR) for the recommended healt...
14        Response (Target)  0 : Customer did not show interest in the reco...
15                      NaN  1 : Customer showed interest in the recommende...

In [3]:
df.head(15)

ID  City_Code  Region_Code  Accomodation_Type  Reco_Insurance_Type  \
0    1         22         3213                  1                    0   
1    2         31         1117                  0                    1   
2    3         31         3732                  0                    0   
3    4         16         4378                  0                    1   
4    5         34         2190                  1                    0   
5    6         35         1785                  1                    0   
6    7         22          679                  0                    0   
7    8          0         3175                  0                    1   
8    9          6         3497                  0                    1   
9   10          0          530                  0                    1   
10  11         20          600                  0                    0   
11  12         19         1097                  0                    1   
12  13         33         3453                  0                    0   
13  14         31          900                  1                    0   
14  15         12         1911                  1                    0   

    Upper_Age  Lower_Age  Is_Spouse  Health Indicator  \
0          36         36          0                 1   
1          75         22          0                 2   
2          32         32          0                 0   
3          52         48          0                 1   
4          44         44          0                 2   
5          52         52          0                 2   
6          28         28          0                 0   
7          75         73          1                 4   
8          52         43          0                 1   
9          59         26          1                 0   
10         21         21          0                 2   
11         59         47          1                 3   
12         66         66          0                 0   
13         20         20          0                 2   
14         27         27          0                 3   

    Holding_Policy_Duration  ...  Reco_Policy_Cat  Reco_Policy_Premium  \
0                      14.0  ...               22              11628.0   
1                       0.0  ...               22              30510.0   
2                       1.0  ...               19               7450.0   
3                      14.0  ...               19              17780.0   
4                       3.0  ...               16              10404.0   
5                       5.0  ...               22              15264.0   
6                       0.0  ...               17              10640.0   
7                       9.0  ...               17              29344.0   
8                      14.0  ...                1              27283.2   
9                       7.0  ...               18              21100.8   
10                      0.0  ...               21               4068.0   
11                      3.0  ...               13              25043.2   
12                      1.0  ...               20              17192.0   
13                      0.0  ...               18               8364.0   
14                      2.0  ...                9               9440.0   

    Response  Cust_Type  Age_Conf_Interval  Reco_Policy_Premium_log  Mean_Age  \
0          0          1                  0                 9.361171      36.0   
1          0          0                 53                10.325810      75.0   
2          1          1                  0                 8.915969      32.0   
3          0          1                  4                 9.785830      52.0   
4          0          1                  0                 9.249946      44.0   
5          1          1                  0                 9.633252      52.0   
6          0          0                  0                 9.272376      28.0   
7          1          1                  2                10.286843      

In [4]:
df.isnull().sum()

ID                            0
City_Code                     0
Region_Code                   0
Accomodation_Type             0
Reco_Insurance_Type           0
Upper_Age                     0
Lower_Age                     0
Is_Spouse                     0
Health Indicator              0
Holding_Policy_Duration       0
Holding_Policy_Type           0
Reco_Policy_Cat               0
Reco_Policy_Premium           0
Response                      0
Cust_Type                     0
Age_Conf_Interval             0
Reco_Policy_Premium_log       0
Mean_Age                      0
Permium_Per_Age_Year          0
Reco_Policy_Premium_Bin_10    0
Reco_Policy_Premium_Bin_50    0
dtype: int64

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50882 entries, 0 to 50881
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   ID                          50882 non-null  int64  
 1   City_Code                   50882 non-null  int64  
 2   Region_Code                 50882 non-null  int64  
 3   Accomodation_Type           50882 non-null  int64  
 4   Reco_Insurance_Type         50882 non-null  int64  
 5   Upper_Age                   50882 non-null  int64  
 6   Lower_Age                   50882 non-null  int64  
 7   Is_Spouse                   50882 non-null  int64  
 8   Health Indicator            50882 non-null  int64  
 9   Holding_Policy_Duration     50882 non-null  float64
 10  Holding_Policy_Type         50882 non-null  float64
 11  Reco_Policy_Cat             50882 non-null  int64  
 12  Reco_Policy_Premium         50882 non-null  float64
 13  Response                    508

In [6]:
df.describe().T

count          mean           std          min  \
ID                          50882.0  25441.500000  14688.512535     1.000000   
City_Code                   50882.0     14.974981     12.147432     0.000000   
Region_Code                 50882.0   1732.788707   1424.081652     1.000000   
Accomodation_Type           50882.0      0.450670      0.497566     0.000000   
Reco_Insurance_Type         50882.0      0.203333      0.402482     0.000000   
Upper_Age                   50882.0     44.856275     17.310271    18.000000   
Lower_Age                   50882.0     42.738866     17.319375    16.000000   
Is_Spouse                   50882.0      0.165520      0.371653     0.000000   
Health Indicator            50882.0      1.882984      1.632672     0.000000   
Holding_Policy_Duration     50882.0      3.621182      4.568238     0.000000   
Holding_Policy_Type         50882.0      1.468417      1.434988     0.000000   
Reco_Policy_Cat             50882.0     15.115188      6.340663     1.000000   
Reco_Policy_Premium         50882.0  14183.950069   6590.074873  2280.000000   
Response                    50882.0      0.239947      0.427055     0.000000   
Cust_Type                   50882.0      0.602001      0.489490     0.000000   
Age_Conf_Interval           50882.0      2.117409      6.865406     0.000000   
Reco_Policy_Premium_log     50882.0      9.447436      0.489185     7.731931   
Mean_Age                    50882.0     44.856275     17.310271    18.000000   
Permium_Per_Age_Year        50882.0      5.724456      0.281541     4.779123   
Reco_Policy_Premium_Bin_10  50882.0      2.397292      1.630808     0.000000   
Reco_Policy_Premium_Bin_50  50882.0     13.992237      8.029032     0.000000   

                                     25%           50%           75%  \
ID                          12721.250000  25441.500000  38161.750000   
City_Code                       4.000000     11.000000     30.000000   
Region_Code                   523.000000   1391.000000   2667.000000   
Accomodation_Type               0.000000      0.000000      1.000000   
Reco_Insurance_Type             0.000000      0.000000      0.000000   
Upper_Age                      28.000000     44.000000     59.000000   
Lower_Age                      27.000000     40.000000     57.000000   
Is_Spouse                       0.000000      0.000000      0.000000   
Health Indicator                1.000000      2.000000      3.000000   
Holding_Policy_Duration         0.000000      2.000000      6.000000   
Holding_Policy_Type             0.000000      1.000000      3.000000   
Reco_Policy_Cat                12.000000     17.000000     20.000000   
Reco_Policy_Premium          9248.000000  13178.000000  18096.000000   
Response                        0.000000      0.000000      0.000000   
Cust_Type                       0.000000      1.000000      1.000000   
Age_Conf_Interval               0.000000      0.000000      0.000000   
Reco_Policy_Premium_log         9.132163      9.486304      9.803446   
Mean_Age                       28.000000     44.000000     59.000000   
Permium_Per_Age_Year            5.520889      5.714061      5.920903   
Reco_Policy_Premium_Bin_10      1.000000      2.000000      3.000000   
Reco_Policy_Premium_Bin_50      8.000000     13.000000     19.000000   

                                     max  
ID                          50882.000000  
City_Code                      35.000000  
Region_Code                  6194.000000  
Accomodation_Type               1.000000  
Reco_Insurance_Type             1.000000  
Upper_Age                      75.000000  
Lower_Age                      75.000000  
Is_Spouse                       1.000000  
Health Indicator                9.000000  
Holding_Policy_Duration        14.000000  
Holding_Policy_Type             4.000000  
Reco_Policy_Cat                22.000000  
Reco_Policy_Premium         43350.400000  
Response                        1.000000  
Cust_Type                       1.00000

#### Split the data into training and testing sets

In [7]:
X = df.drop('Response', axis=1)
y = df['Response']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

X_train.shape, X_test.shape, y_train.shape, y_test.shape

((40705, 20), (10177, 20), (40705,), (10177,))

In [8]:
#Scaling the data
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [9]:
#Importing Machine Learning Models
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score
from sklearn.model_selection import validation_curve

In [10]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
C_param_range = [0.001, 0.01, 0.1, 1, 10, 100]

t = pd.DataFrame(columns=['C_parameter', 'Accuracy'])
t['C_parameter'] = C_param_range

for j, C in enumerate(C_param_range):
    Logreg = LogisticRegression(penalty='l2', C=C, random_state=40)
    Logreg.fit(X_train, y_train)
    y_pred_lr = Logreg.predict(X_test)
    t.at[j, 'Accuracy'] = accuracy_score(y_test, y_pred_lr)

t

C_parameter  Accuracy
0        0.001  0.755429
1        0.010  0.755429
2        0.100  0.755429
3        1.000  0.755429
4       10.000  0.755429
5      100.000  0.755429

##### The results indicate that the accuracy of the logistic regression model remains the same across all tested values of the C parameter.


In [11]:
from sklearn.metrics import confusion_matrix
cnf_matrix= confusion_matrix(y_test,y_pred_lr)
print(cnf_matrix)
Accuracy_lr=Logreg.score(X_test,y_test)
print(Accuracy_lr)

[[7688    0]
 [2489    0]]
0.7554289083226884


In [12]:
from sklearn.model_selection import cross_val_score

cv_scores_test= cross_val_score(Logreg,X_test,y_test,cv=5,scoring='roc_auc')
cv_scores_train= cross_val_score(Logreg,X_train,y_train,cv=5,scoring='roc_auc')
print(cv_scores_test)
cv_scores_lr_test= cv_scores_test.mean()
cv_scores_lr_train= cv_scores_train.mean()
cv_scores_std_test_lr= cv_scores_test.std()
print ('Mean cross validation test score: ' +str(cv_scores_lr_test))
print ('Mean cross validation train score: ' +str(cv_scores_lr_train))
print ('Standard deviation in cv test scores: ' +str(cv_scores_std_test_lr))

[0.54516767 0.54856618 0.56189543 0.57964192 0.55239697]
Mean cross validation test score: 0.557533632720381
Mean cross validation train score: 0.5645633656534988
Standard deviation in cv test scores: 0.01238966982340089


In [13]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score, roc_auc_score

knn = KNeighborsClassifier(p=2, weights='distance', n_neighbors=50)
knn.fit(X_train, y_train)

y_predict_knn = knn.predict(X_test)

cnf_matrix = confusion_matrix(y_test, y_predict_knn)
print("Confusion Matrix:")
print(cnf_matrix)

accuracy_knn = accuracy_score(y_test, y_predict_knn)
print("Accuracy:", accuracy_knn)

precision_knn = precision_score(y_test, y_predict_knn)
recall_knn = recall_score(y_test, y_predict_knn)
f1_knn = f1_score(y_test, y_predict_knn)
roc_auc_knn = roc_auc_score(y_test, y_predict_knn)

print("Precision:", precision_knn)
print("Recall:", recall_knn)
print("F1 Score:", f1_knn)
print("AUC-ROC:", roc_auc_knn)

Confusion Matrix:
[[7687    1]
 [2489    0]]
Accuracy: 0.7553306475385674
Precision: 0.0
Recall: 0.0
F1 Score: 0.0
AUC-ROC: 0.4999349635796046


In [16]:
cv_scores_test= cross_val_score(knn,X_test,y_test,cv=5,scoring='roc_auc')
cv_scores_train= cross_val_score(knn,X_train,y_train,cv=5,scoring='roc_auc')
print(cv_scores_test)
cv_scores_knn_test= cv_scores_test.mean()
cv_scores_knn_train= cv_scores_train.mean()
cv_scores_std_knn= cv_scores_test.std()
print ('Mean cross validation test score: ' +str(cv_scores_knn_test))
print ('Mean cross validation train score: ' +str(cv_scores_knn_train))
print ('Standard deviation in cv scores: ' +str(cv_scores_std_knn))

[0.53527635 0.55010549 0.52287195 0.57396012 0.54183161]
Mean cross validation test score: 0.544809104126112
Mean cross validation train score: 0.546635020529647
Standard deviation in cv scores: 0.01708104891183556


In [17]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(bootstrap=True,n_estimators=100,criterion='entropy')
rf.fit(X_train, y_train)
y_predict_rf = rf.predict(X_test)

#Confusion matrix:
cnf_matrix = confusion_matrix(y_test, y_predict_rf)
print(cnf_matrix)
Accuracy_rf=rf.score(X_test,y_test)
print(Accuracy_rf)

[[7642   46]
 [2466   23]]
0.7531689102879041


In [18]:
cv_scores_test= cross_val_score(rf,X_test,y_test,cv=5,scoring='roc_auc')
cv_scores_train= cross_val_score(rf,X_train,y_train,cv=5,scoring='roc_auc')
print(cv_scores_test)
cv_scores_rf_test= cv_scores_test.mean()
cv_scores_rf_train= cv_scores_train.mean()
cv_scores_std_rf= cv_scores_test.std()
print ('Mean cross validation test score: ' +str(cv_scores_rf_test))
print ('Mean cross validation train score: ' +str(cv_scores_rf_train))
print ('Standard deviation in cv scores: ' +str(cv_scores_std_rf))

[0.56421917 0.56864519 0.56363735 0.59949362 0.57715181]
Mean cross validation test score: 0.5746294263311824
Mean cross validation train score: 0.6040969312776461
Standard deviation in cv scores: 0.013338670285312468


In [19]:
from sklearn.svm import SVC

svm = SVC(kernel='linear')
svm.fit(X_train, y_train)

# Predict using model:

y_predict_svm=svm.predict(X_test)

#Confusion matrix:

cnf_matrix = confusion_matrix(y_test, y_predict_svm)
print(cnf_matrix)

Accuracy_svm=svm.score(X_test,y_test)
print(Accuracy_svm)

[[7688    0]
 [2489    0]]
0.7554289083226884


In [20]:
cv_scores_test= cross_val_score(svm,X_test,y_test,cv=5,scoring='roc_auc')
cv_scores_train= cross_val_score(svm,X_train,y_train,cv=5,scoring='roc_auc')
print(cv_scores_test)
cv_scores_svm_test= cv_scores_test.mean()
cv_scores_svm_train= cv_scores_train.mean()
cv_scores_std_svm= cv_scores_test.std()
print ('Mean cross validation test score: ' +str(cv_scores_svm_test))
print ('Mean cross validation train score: ' +str(cv_scores_svm_train))
print ('Standard deviation in cv scores: ' +str(cv_scores_std_svm))

[0.51180274 0.51336033 0.49529426 0.50137309 0.53298033]
Mean cross validation test score: 0.5109621504426171
Mean cross validation train score: 0.5136752749737526
Standard deviation in cv scores: 0.01287393295055191


In [21]:
from sklearn.ensemble import ExtraTreesClassifier, GradientBoostingClassifier
gbc = GradientBoostingClassifier(subsample=0.8, learning_rate=0.05 , n_estimators=160, random_state=5, max_depth=9, max_leaf_nodes=100)
gbc.fit(X_train, y_train)

y_predict_gbc = gbc.predict(X_test)

cnf_matrix = confusion_matrix(y_test, y_predict_gbc)
print(cnf_matrix)
Accuracy_gbc=gbc.score(X_test,y_test)
print(Accuracy_gbc)

[[7613   75]
 [2426   63]]
0.7542497789132357


In [22]:
cv_scores_test= cross_val_score(gbc,X_test,y_test,cv=5,scoring='roc_auc')
cv_scores_train= cross_val_score(gbc,X_train,y_train,cv=5,scoring='roc_auc')
print(cv_scores_test)
cv_scores_gbc_test= cv_scores_test.mean()
cv_scores_gbc_train= cv_scores_train.mean()
cv_scores_std_gbc= cv_scores_test.std()
print ('Mean cross validation test score: ' +str(cv_scores_gbc_test))
print ('Mean cross validation train score: ' +str(cv_scores_gbc_train))
print ('Standard deviation in cv scores: ' +str(cv_scores_std_gbc))

[0.58815365 0.60290447 0.57461937 0.62978002 0.59212386]
Mean cross validation test score: 0.5975162738653578
Mean cross validation train score: 0.6519043463882364
Standard deviation in cv scores: 0.018498577413988974


In [29]:
classifiers = [Logreg, knn, svm, rf, gbc]

cv_scores_test = {}
cv_scores_train = {}

for clf in classifiers:
    clf_name = clf.__class__.__name__
    cv_scores_test[clf_name] = cross_val_score(clf, X_test, y_test, cv=5, scoring='roc_auc').mean()
    cv_scores_train[clf_name] = cross_val_score(clf, X_train, y_train, cv=5, scoring='roc_auc').mean()

print("Cross-Validation AUC-ROC Scores (Test):")
print(cv_scores_test)
print("Cross-Validation AUC-ROC Scores (Train):")
print(cv_scores_train)

accuracy_scores = {}
for clf in classifiers:
    clf_name = clf.__class__.__name__
    clf.fit(X_train, y_train)
    accuracy_scores[clf_name] = clf.score(X_test, y_test)

print("Accuracy Scores:")
print(accuracy_scores)

Cross-Validation AUC-ROC Scores (Test):
{'LogisticRegression': 0.557533632720381, 'KNeighborsClassifier': 0.544809104126112, 'SVC': 0.5109621504426171, 'RandomForestClassifier': 0.5714719162576385, 'GradientBoostingClassifier': 0.5975162738653578}
Cross-Validation AUC-ROC Scores (Train):
{'LogisticRegression': 0.5645633656534988, 'KNeighborsClassifier': 0.546635020529647, 'SVC': 0.5136752749737526, 'RandomForestClassifier': 0.6015457416339116, 'GradientBoostingClassifier': 0.6519043463882364}
Accuracy Scores:
{'LogisticRegression': 0.7554289083226884, 'KNeighborsClassifier': 0.7553306475385674, 'SVC': 0.7554289083226884, 'RandomForestClassifier': 0.7528741279355409, 'GradientBoostingClassifier': 0.7542497789132357}


In [31]:
results_df = pd.DataFrame({
    'Model': ['Logistic Regression', 'KNN', 'SVM', 'Random Forest', 'Gradient Boost'],
    'CV Test Score': [cv_scores_test['LogisticRegression'], cv_scores_test['KNeighborsClassifier'], cv_scores_test['SVC'], cv_scores_test['RandomForestClassifier'], cv_scores_test['GradientBoostingClassifier']],
    'CV Train Score': [cv_scores_train['LogisticRegression'], cv_scores_train['KNeighborsClassifier'], cv_scores_train['SVC'], cv_scores_train['RandomForestClassifier'], cv_scores_train['GradientBoostingClassifier']],
    'Accuracy': [accuracy_scores['LogisticRegression'], accuracy_scores['KNeighborsClassifier'], accuracy_scores['SVC'], accuracy_scores['RandomForestClassifier'], accuracy_scores['GradientBoostingClassifier']]
})

print(results_df)

                 Model  CV Test Score  CV Train Score  Accuracy
0  Logistic Regression       0.557534        0.564563  0.755429
1                  KNN       0.544809        0.546635  0.755331
2                  SVM       0.510962        0.513675  0.755429
3        Random Forest       0.571472        0.601546  0.752874
4       Gradient Boost       0.597516        0.651904  0.754250


#### Observation

###### 1) Gradient Boosting has the highest CV Test Score (0.597516) and CV Train Score (0.651904), indicating it performed well during cross-validation, though it might be slightly overfitting given the train score. 
###### 2) Random Forest has a good CV Test Score (0.571472) but shows a larger difference between train (0.601546) and test scores, suggesting potential overfitting.

In [33]:
# Gradient Boosting Parameter Grid
param_grid_gb = {
    'learning_rate': [0.01, 0.1, 0.05],
    'n_estimators': [100, 200, 300],
    'max_depth': [3, 5, 7],
    'subsample': [0.8, 0.9, 1.0]
}

# Random Forest Parameter Grid
param_grid_rf = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}



In [ ]:
# Gradient Boosting Grid Search
gb_model = GradientBoostingClassifier(random_state=42)
grid_search_gb = GridSearchCV(estimator=gb_model, param_grid=param_grid_gb, cv=5, n_jobs=-1, verbose=2)
grid_search_gb.fit(X_train, y_train)


Fitting 5 folds for each of 81 candidates, totalling 405 fits


In [ ]:
rf_model = RandomForestClassifier(random_state=42)
grid_search_rf = GridSearchCV(estimator=rf_model, param_grid=param_grid_rf, cv=5, n_jobs=-1, verbose=2)
grid_search_rf.fit(X_train, y_train)
